Realm/RLMLogger.mm

```diff
@@ -0,0 +1,192 @@
+        case Level::all:    return RLMLogLevelAll;
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+static NSString* levelPrefix(Level logLevel) {
+    switch (logLevel) {
+        case Level::off:
+        case Level::all:    return @"";
+        case Level::trace:  return @"Trace";
+        case Level::debug:  return @"Debug";
+        case Level::detail: return @"Detail";
+        case Level::info:   return @"Info";
+        case Level::error:  return @"Error";
+        case Level::warn:   return @"Warning";
+        case Level::fatal:  return @"Fatal";
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+struct CocoaLogger : public Logger {
+    void do_log(Level level, const std::string& message) override {
+        NSLog(@"%@: %@", levelPrefix(level), RLMStringDataToNSString(message));
+    }
+};
+
+class CustomLogger : public Logger {
+public:
+    RLMLoggerFunction function;
+    void do_log(Level level, const std::string& message) override {
+        @autoreleasepool {
+            if (function) {
+                function(logLevelForLevel(level), RLMStringDataToNSString(message));
+            }
+        }
+    }
+};
+
+struct LoggerFunctionCallbackWrapper {
+    void (^block)(RLMLogLevel, NSString *);
+};
+
+void RLMSetDefaultLogLevel(Level level) {
+    Logger::set_default_level_threshold(level);
+}
+} // anonymous namespace
+
+@implementation RLMLogger {
+    std::shared_ptr<Logger> _logger;
+}
+
+static RLMLogger *_defaultLogger;
+static RLMLogLevel _defaultLogLevel = RLMLogLevelWarn;
+
+- (RLMLogLevel)level {
+    return logLevelForLevel(_logger->get_level_threshold());
+}
+
+- (void)setLevel:(RLMLogLevel)logLevel {
+    _logger->set_level_threshold(levelForLogLevel(logLevel));
+}
+
+- (instancetype)initWithLogger:(std::shared_ptr<Logger>)logger {
+    if (self = [self init]) {
+        self->_logger = logger;
+    }
+    return self;
+}
+
+- (instancetype)init {
+    if (self = [super init]) {
+        auto logger = std::make_shared<CustomLogger>();
+        auto wrapper = LoggerFunctionCallbackWrapper{(^(RLMLogLevel level, NSString *message) {
+            [self doLog:level message:message];
+        })};
+        logger->function = wrapper.block;
+        self->_logger = logger;
+    }
+    return self;
+}
+
+- (void)doLog:(RLMLogLevel)logLevel message:(NSString *)message {}
+
+- (void)logWithLevel:(RLMLogLevel)logLevel message:(NSString *)message, ... {
+    auto level = levelForLogLevel(logLevel);
+    if (_logger->would_log(level)) {
+        va_list args;
+        va_start(args, message);
+        _logger->log(level, message.UTF8String, args);
+        va_end(args);
+    }
+}
+
+- (void)logLevel:(RLMLogLevel)logLevel message:(NSString *)message {
+    auto level = levelForLogLevel(logLevel);
+    if (_logger->would_log(level)) {
+        _logger->log(level, message.UTF8String);
+    }
+}
+
+#pragma mark Global Logger Setter
+
++ (RLMLogLevel)logLevel {
+    return _defaultLogLevel;
```

Comment:
This should be `Logger::get_default_level_threshold()` rather than having a separate duplicated value.